ext/filter/logical_filters.c

```diff
@@ -796,35 +796,67 @@ static int _php_filter_validate_ipv6(char *str, size_t str_len) /* {{{ */
 				return 0;
 			}
 			if (*str == ':') {
-				if (compressed) {
+				if (compressed_pos >= 0) {
 					return 0;
 				}
-				blocks++; /* :: means 1 or more 16-bit 0 blocks */
-				compressed = 1;
-
+				if (ip && blocks < 8) {
+					ip[blocks] = -1;
+				}
+				compressed_pos = blocks++; /* :: means 1 or more 16-bit 0 blocks */
 				if (++str == end) {
-					return (blocks <= 8);
+					if (blocks > 8) {
+						return 0;
+					}
+					goto fixup_ip;
 				}
 			} else if ((str - 1) == s) {
 				/* don't allow leading : without another : following */
 				return 0;
 			}
 		}
-		n = 0;
-		while ((str < end) &&
-		       ((*str >= '0' && *str <= '9') ||
-		        (*str >= 'a' && *str <= 'f') ||
-		        (*str >= 'A' && *str <= 'F'))) {
+		num = n = 0;
+		while (str < end) {
+			if (*str >= '0' && *str <= '9') {
+				num = 16 * num + (*str - '0');
+			} else if (*str >= 'a' && *str <= 'f') {
+				num = 16 * num + (*str - 'a') + 10;
+			} else if (*str >= 'A' && *str <= 'F') {
+				num = 16 * num + (*str - 'A') + 10;
+			} else {
+				break;
+			}
 			n++;
 			str++;
 		}
+		if (ip && blocks < 8) {
+			ip[blocks] = num;
+		}
 		if (n < 1 || n > 4) {
 			return 0;
 		}
 		if (++blocks > 8)
 			return 0;
 	}
-	return ((compressed && blocks <= 8) || blocks == 8);
+
+fixup_ip:
+	if (ip && ipv4) {
+		for (i = 0; i < 5; i++) {
+			ip[i] = 0;
+		}
+		ip[i++] = 0xffff;
+		ip[i++] = 256 * ip4elm[0] + ip4elm[1];
+		ip[i++] = 256 * ip4elm[2] + ip4elm[3];
+	} else if (ip && compressed_pos >= 0 && blocks <= 8) {
+		int offset = 8 - blocks;
+		for (i = 7; i > compressed_pos + 1; i--) {
```

Comment:
Hm, is this condition right? That meas the lowest element copied is from `ip[compressed_pos - offset + 2]` which doesn't look correct.

Reply:
If only one element is compressed, that loop copies elements to itself. If more than one element is compressed, the element of `compressed_pos+1` won't need to be copied anyway. I agree, though, that this is a confusing optimization.

Reply:
I still don't think this is right... Let's consider `0:1::2` for example. Here we would have block=3, compressed_pos=2, offset =5 right? So the move would result in something like this:

```
// before move
0 1 2 u u u u u
// after move
0 1 x x x 0 1 2
```
Where `x` has been read from before the start of the buffer. Am I mistunderstand something?

Reply:
For this example (`0:1::2`) we would have `blocks==4` (the `compressed_pos` is counted as block), `compressed_pos==2` and `offset==4`. The ip array would be `0 1 -1 2 u u u u`. Then element 7 to 3 would be set to the values element 3 to -1 (so the +1 isn't an optimization, actually; I need to revert that); with the previous version element 7 to 4 would be set to the values of element 3 to 0. This would result in `0 1 -1 2 0 1 -1 2`. The next `for` loop will set the elements 6 to 2 to `0`, resulting in `0 1 0 0 0 0 0 2`, which is the desired result.

So the previous version was correct, but did unnecessary copying.

Reply:
I think it would be most obvious if you use `i > compressed_pos + offset` as the limit, because that's where the second loop that does the zeroing continues from.

Reply:
D'oh